Fix crypto/keys get_private_ECDSA to take private exponent hex

timchurchard opened the pull request.

StephanieBracaloni commented on iotics/lib/identity/api/advanced_api.py:
```diff
@@ -24,6 +24,22 @@
 
 class AdvancedIdentityLocalApi:
+    @staticmethod
+    def get_key_pair_from_hex_private_key(private_expo_hex: str) -> KeyPair:
```
:+1:

StephanieBracaloni commented on iotics/lib/identity/register/document_builder.py:
```diff
@@ -242,11 +246,11 @@ def build_from_dict(self, data: dict) -> RegisterDocument:
         try:
             for k in data['publicKey']:
                 self.add_public_key_obj(RegisterPublicKey.from_dict(k))
-            for k in data['authentication']:
```
:+1:

StephanieBracaloni commented on iotics/lib/identity/register/document_builder.py:
```diff
@@ -28,7 +28,11 @@ def __init__(self):
     def _check_and_update_map(self, name: str, val: object, key_map: dict, message_prefix: str):
         if name in self.name_set:
-            raise IdentityRegisterDocumentKeyConflictError(f'{message_prefix} \'{name}\' already in use')
+            if isinstance(val, RegisterDelegationProof) and key_map[name].controller == val.controller:
```
What are we trying to achieve here?

timchurchard replied:
I found that if you call `create_twin_with_control_delegation` twice it would raise this error

StephanieBracaloni replied:
twice with the same name?
Maybe  we could improve the check so it raises only if the new key is different from the existing one but keep it generic:

```
if name in self.name_set and key_map.get(name) != val:
```

timchurchard replied:
Cool thanks I'll update the check

StephanieBracaloni commented on tests/unit/iotics/lib/identity/register/test_document_builder.py:
```diff
@@ -258,11 +275,39 @@ def test_can_remove_not_existing_key_without_error(doc_did, doc_proof, doc_keys,
     assert existing_doc_key_set == new_doc_key_set
 
 
+def test_building_a_register_doc_does_not_raises_if_double_add(doc_keys):
```
can this be renamed to
`test_building_a_register_doc_does_not_raises_if_same_key_is_added_twice`?

StephanieBracaloni commented on tests/unit/iotics/lib/identity/register/test_document_builder.py:
```diff
@@ -258,11 +275,39 @@ def test_can_remove_not_existing_key_without_error(doc_did, doc_proof, doc_keys,
+            .add_public_key_obj(duplicate_pub_key1)
+
+
+def test_building_a_register_doc_does_not_raises_if_double_add_delegation(doc_keys):
```
same than above

StephanieBracaloni commented on tests/unit/iotics/lib/identity/register/test_document_builder.py:
```diff
@@ -134,6 +134,23 @@ def test_can_build_a_register_doc_from_dict(minimal_doc, full_doc, is_minimal):
 
 
+@pytest.mark.parametrize('is_minimal', (True, False))
+def test_can_build_a_register_doc_from_dict_absent_auth(minimal_doc, full_doc, is_minimal):
```
this one is not testing the "absence of auth_keys or public keys or delegation" fields.
Because to_dict() will set those fields with an explicit empty list

This test could be:
```
def test_can_build_a_register_doc_from_minimal_dict(minimal_doc):
    doc_as_dict = minimal_doc.to_dict()
    doc_as_dict.pop('authentication')
    doc_as_dict.pop('delegateControl')
    doc_as_dict.pop('delegateAuthentication')

    new_doc = RegisterDocumentBuilder().build_from_dict(doc_as_dict)
    assert new_doc.did == doc.did
    assert new_doc.purpose == doc.purpose
    assert new_doc.proof == doc.proof
    assert new_doc.revoked == doc.revoked
    assert new_doc.spec_version == doc.spec_version
    assert new_doc.metadata == doc.metadata
    assert new_doc.public_keys == doc.public_keys
    assert not new_doc.auth_keys
    assert not new_doc.control_delegation_proof 
    assert not new_doc.auth_delegation_proof 
```

Commit:
Fix crypto/keys get_private_ECDSA to take private exponent hex

- Change document_builder to not raise DocumentKeyConflictError if key already exists
- Change document_builder to allow authentication, delegateAuthentication and delegateControl to be absent
- Change advanced_api create token functions to take KeyPair instead of KeyPairSecrets
- Add is_equal to RegisterKey and RegisterDelegationProof

StephanieBracaloni reviewed (approved):
LGTM :+1:
